### ISSUE TITLE
Retirar imagem dos objetos, detecção esta a falhar

### ISSUE THREAD

**Issue by hugohashi** [miguelriemoliveira/savi_23-24#11]
Olá, professor, boa noite.

Temos algumas dúvidas no trabalho.

1º - Como podemos retirar a iimagem dos objetos através da nuvem?

2º - Conseguimos treinar o modelo e fizemos testes de métricas que deram relativamente bem, mas em real-time ele classifica os objetos mal mesmo apresentando niveis de confiança elevados.

**Comment by miguelriemoliveira** [comment]
Olá @hugohashi 

> Como podemos retirar a imagem dos objetos através da nuvem?

Para isso terão de ter uma imagem e uma nuvem de pontos registadas, quer dizer, em que seja possível ter uma transformação geométrica ou um mapeamento entre elas.

No dataset há nuvens de pontos e imagens correspondentes? Envie um exemplo pff.

> 2ª Conseguimos treinar o modelo e fizemos testes de métricas que deram relativamente bem, mas em real-time ele classifica os objetos mal mesmo apresentando niveis de confiança elevados.

Pelas métricas realmente parecem valores de Precision e Recall muito bons, por isso parece que o treino correu muito bem. Está a usar imagens diferentes para treinar / validar e testar?

Quando diz tempo real, é um programa que adquire a imagem da câmara e deteta os objetos que vê?
Descreva a experiência com mais detalhe para eu tentar perceber.

**Comment by ritapm18** [comment]
Olá professor, penso que não temos correspondência entre imagens e nuvem de pontos... pois só estamos a ler a nuvem de pontos de cada cena e nada mais... Deveriamos ler as imagens e torna-las em nuvem e depois conforme a deteção é que selecionamos a imagem, é isso? Ou seja não há maneira de através de uma nuvem de pontos criar uma imagem, mas há maneira de criar uma nuvem de pontos através de várias imagens.

**Comment by hugohashi** [comment]
Boa noite, professor @miguelriemoliveira 

> Pelas métricas realmente parecem valores de Precision e Recall muito bons, por isso parece que o treino correu muito bem. Está a usar imagens diferentes para treinar / validar e testar?


Sim, usamos o código de split data que o senhor deu em uma das aulas - separamos o dataset em treino e teste (80%/20%) (pulamos a parte de validation, ja que o enunciado diz para separarmos assim) e depois salvamos em um arquivo json

> Quando diz tempo real, é um programa que adquire a imagem da câmara e deteta os objetos que vê? Descreva a experiência com mais detalhe para eu tentar perceber.

Sim fazemos a transformacao do frame lido pela camera (resize + to tensor), o unsqueeze e usamos como input para o modelo.

Após fazer uma pesquisa melhor sobre este problema comecei a pensar que pode ser um caso de overfitting. Durante a fase de treino colocamos um if statement para parar o treinamento do modelo (usando valores que pareciam bons para o epoch_training_loss e para o epoch_testing_loss ), mas acho que não foi a melhor maneira de fazer isso.

Aqui segue o código do opencv:

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

# Load the trained model
model = Model().to(device)
checkpoint = torch.load('models/model.pkl', map_location=device)
model.load_state_dict(checkpoint['model_state_dict'])

model.eval()

transform = transforms.Compose([
    transforms.ToPILImage(),
    transforms.Resize((224, 224)),
    transforms.ToTensor()
])

# Open the video capture (0 for webcam)
cap = cv2.VideoCapture(0)

with open('json_files/label_mapping.json', 'r') as f:
    label_mapping = json.load(f)

labels = list(label_mapping.keys())

while True:
    # Read a frame from the video capture
    ret, frame = cap.read()

    height, width, _ = frame.shape

    frame = cv2.flip(frame, 1)

    obj_detection = copy.deepcopy(frame)

    # Preprocess the input frame
    input_tensor = transform(obj_detection)
    input_tensor = input_tensor.to(device)
    input_batch = input_tensor.unsqueeze(0)

    # Perform inference
    with torch.no_grad():
        detections = model(input_batch)

    # Post-process and draw bounding boxes
    probabilities = F.softmax(detections, dim=1)

    _, class_indices = torch.max(probabilities, 1)
    
    for i, detection in enumerate(detections):
        confidence = probabilities[i, class_indices[i]].item()
        class_id = class_indices[i].item()
        class_name = labels[class_id]
        
        # TODO: Extract bounding box coordinates
        
        if confidence > 0.8:
        #     cv2.rectangle(frame, (x, y), (x + w, y + h), (0, 255, 0), 2)
            cv2.putText(obj_detection, f"Object detected: {class_name}", (25, 50), cv2.FONT_HERSHEY_SIMPLEX, 0.8, (0, 255, 0), 2)
            cv2.putText(obj_detection, f"With confidence of: {confidence}", (25, 100), cv2.FONT_HERSHEY_SIMPLEX, 0.8, (0, 255, 0), 2)


    # Display the frame with bounding boxes
    cv2.imshow('Original frame', frame)
    cv2.imshow('Object Detection', obj_detection)

    # Break the loop if 'q' key is pressed
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

# Release the video capture and close the OpenCV windows
cap.release()
cv2.destroyAllWindows()


E aqui está o if statement que foi adicionado ao código do Trainer (que também usou o do professor como base) para parar o treinamento do modelo:
            #Stop training the model if both losses are smaller than 0.1
            self.saveModel(model=self.model,
                        optimizer=self.optimizer,
                        epoch_idx=epoch_idx,
                        epoch_train_losses=epoch_train_losses,
                        epoch_test_losses=epoch_test_losses)
            print("model saved")

            self.draw(epoch_train_losses, epoch_test_losses)

            if epoch_train_loss < 0.1 and epoch_test_loss < 0.1:
                break

**Comment by miguelriemoliveira** [comment]
> Olá professor, penso que não temos correspondência entre imagens e nuvem de pontos... pois só estamos a ler a nuvem de pontos de cada cena e nada mais... Deveriamos ler as imagens e torna-las em nuvem e depois conforme a deteção é que selecionamos a imagem, é isso? Ou seja não há maneira de através de uma nuvem de pontos criar uma imagem, mas há maneira de criar uma nuvem de pontos através de várias imagens.

Olá @ritapm18 ,

então não tem para uma determinada nuvem de pontos a imagem que lhe corresponde? Tem a certeza? Porque isso iria ajudar muito. 

Se assim não for acho difícil fazer a conversão de que fala. Quero dizer, temos um conjunto de pontos na nuvem de pontos a que pertence o objeto, mas não é possível (não é fácil e o resultado é mau) criar uma imagem partindo da nuvem de pontos.

**Comment by miguelriemoliveira** [comment]
Olá @hugohashi ,

não vejo nada de mal no seu código, mas também não estive a ver com grande detalhe.

O overfitting é normal, com o reconhecimento em tempo real com objetos ligeiramente diferentes dos treinados é normal que o precision e o recall baixem, mas não é normal que não consiga reconhecer nenhum objeto bem.

Talvez as imagens que usou para treinar sejam muito diferentes das que está a usar para reconhecer em tempo real. Qual é a resolução de cada uma delas?

**Comment by ritapm18** [comment]
Já vi que no dataset das cenas existem imagens para cada cena, é suposto fazer a nuvem através disso? 
Por aqui: open3d.geometry.create_point_cloud_from_rgbd_image(image, intrinsic, extrinsic=(with default value)) ?
Eu estava apenas ler as nuvens que já existiam no dataset e queria extrair a imagem daí... mas então não dá para fazer isso.

**Comment by hugohashi** [comment]
Boa noite, professor @miguelriemoliveira , 

> Olá @hugohashi ,
> 
> não vejo nada de mal no seu código, mas também não estive a ver com grande detalhe.
> 
> O overfitting é normal, com o reconhecimento em tempo real com objetos ligeiramente diferentes dos treinados é normal que o precision e o recall baixem, mas não é normal que não consiga reconhecer nenhum objeto bem.
> 
> Talvez as imagens que usou para treinar sejam muito diferentes das que está a usar para reconhecer em tempo real. Qual é a resolução de cada uma delas?

Como assim resolução? Se for o tamanho eu apliquei o resize (224,224) como o senhor fez na aula.

Se calhar (e se não for um problema pro professor) é melhor marcarmos uma reunião para perceber melhor nossas dúvidas, e não estar a spamar o issues.

**Comment by miguelriemoliveira** [comment]
> Já vi que no dataset das cenas existem imagens para cada cena, é suposto fazer a nuvem através disso?
> Por aqui: open3d.geometry.create_point_cloud_from_rgbd_image(image, intrinsic, extrinsic=(with default value)) ?
> Eu estava apenas ler as nuvens que já existiam no dataset e queria extrair a imagem daí... mas então não dá para fazer isso.

A minha ideia era encontrar no dataset as imagens e as nuvens de pontos que lhes correspondem. Se isso não é possível, sim, pode usar essa função para obter as nuvens de pontos a partir de uma imagem rgb-d. Mas note que uma imagem rgb-d não é uma imagem rgb ... você tem imagens rgb-d no dataset?

**Comment by miguelriemoliveira** [comment]
> Como assim resolução? Se for o tamanho eu apliquei o resize (224,224) como o senhor fez na aula.

Resolução das imagens antes de fazer o resize. Qual é a das imagens do dataset, e qual é a da sua câmara?

Podemos marcar uma reunião, sim, mas os issues são para isto mesmo. Para discutir e tirar dúvidas.

Eu consigo na próxima segunda feira ao final da manhã. Por vocês é possível?

**Comment by ritapm18** [comment]
Olá professor, tanto eu como o Hugo trabalhamos e não conseguimos reunir ao final da manhã. Teria de ser ao final do dia (+- 19h).

**Comment by ritapm18** [comment]
E realmente não há imagens RGBD no dataset, apenas RGB e depth...

**Comment by miguelriemoliveira** [comment]
> Olá professor, tanto eu como o Hugo trabalhamos e não conseguimos reunir ao final da manhã. Teria de ser ao final do dia (+- 19h).

Ok, pode ser hoje pelas 21h30?

**Comment by ritapm18** [comment]
Por mim sim :)

**Comment by miguelriemoliveira** [comment]
Ok, esperamos a resposta do @hugohashi ...

**Comment by hugohashi** [comment]
@miguelriemoliveira  boa tarde professor por mim pode ser esse horario sim

**Comment by miguelriemoliveira** [comment]
Boa noite @hugohashi e @ritapm18 . Vou  agora abrir a sessão zoom de savi.

Até já.
